File: src/2023/day-10/main.ts
import { readLines } from "../../util/string";

type Position = [number, number];
const NORTH: Position = [-1, 0],
    EAST: Position = [0, 1],
    SOUTH: Position = [1, 0],
    WEST: Position = [0, -1];

const directionMap = {
    "|": [NORTH, SOUTH],
    "-": [EAST, WEST],
    L: [NORTH, EAST],
    J: [NORTH, WEST],
    "7": [SOUTH, WEST],
    F: [SOUTH, EAST],
    ".": [],
    S: [NORTH, EAST, SOUTH, WEST],
} as Record<string, Position[]>;

// const grid = parseInput("example.txt");
const grid = readLines("input.txt");
console.warn("part one answer", getPartOneAnswer(grid));

// const grid = parseInput("example.txt");
// const grid = parseInput("input.txt");
// console.warn("part two answer", getPartOneAnswer(grid));

function getPartOneAnswer(grid: string[]) {
    let startPosition: Position = findChar(grid, "S")!;
    const { cycleLen } = dfs(startPosition, grid, { cycleLen: 1 });
    return Math.floor(cycleLen / 2);
}

function dfs(start: Position, grid: string[], shared: { cycleLen: number }) {
    const stack = [
        {
            node: start,
            parent: [-1, -1] as Position,
            pathLen: 0,
            childIndex: -1,
        },
    ];
    const doneVisiting = new Map<string, boolean>([[`${start}`, false]]);
    while (stack.length > 0) {
        const { node, parent, pathLen, childIndex } = stack.pop()!;
        const [i, j] = node;
        doneVisiting.set(`${node}`, false);
        const nextPathLen = pathLen + 1;
        let neighIndex = 0;
        for (const [di, dj] of directionMap[grid[i][j]]) {
            const neigh: Position = [i + di, j + dj];
            const outsideOrParent =
                grid[neigh[0]]?.[neigh[1]] == null ||
                `${neigh}` === `${parent}`;
            if (outsideOrParent) continue;
            const done = doneVisiting.get(`${neigh}`);
            if (done == null) {
                stack.push({
                    node: neigh,
                    parent: node,
                    pathLen: nextPathLen,
                    childIndex: neighIndex++,
                });
            } else if (done === false) {
                shared.cycleLen = Math.max(shared.cycleLen, nextPathLen);
                break;
            }
        }
        if (childIndex === 0) {
            doneVisiting.set(`${parent}`, true);
        }
    }
    return shared;
}

function findChar(grid: string[], char: string) {
    for (let i = 0; i < grid.length; i++) {
        for (let j = 0; j < grid[0].length; j++) {
            if (grid[i][j] === char) {
                return [i, j] as Position;
            }
        }
    }
    return null;
}

function getPartTwoAnswer(grid: string[]) {}

File: src/2023/day-10/main.py
import sys
from math import floor

new_limit = 500_000_000  # Set your desired recursion limit here
sys.setrecursionlimit(new_limit)

def read_lines(file_name):
    with open(file_name) as file:
        return list(map(str.strip, file.readlines()))

# Set the new recursion limit



    
NORTH = (-1, 0)
EAST = (0, 1)
SOUTH = (1, 0)
WEST = (0, -1)

direction_map = {
    "|": [NORTH, SOUTH],
    "-": [EAST, WEST],
    "L": [NORTH, EAST],
    "J": [NORTH, WEST],
    "7": [SOUTH, WEST],
    "F": [SOUTH, EAST],
    ".": [],
    "S": [NORTH, EAST, SOUTH, WEST],
}

def find_char(grid, char):
    for i, row in enumerate(grid):
        for j, c in enumerate(row):
            if c == char:
                return i, j

def part_one(grid):
    animal_pos = find_char(grid, "S")
    done_visiting = {}
    cycle_len = 0

    def dfs(start, parent, path_len):
        nonlocal cycle_len
        i, j = start
        outside = not (0 <= i < len(grid) and 0 <= j < len(grid[0]))
        done = done_visiting.get(start, None)
        if outside or (done is True):
            return
        if done is False:
            cycle_len = max(cycle_len, path_len)
            return
        done_visiting[start] = False
        for di, dj in direction_map[grid[i][j]]:
            neigh = (i + di, j + dj)
            if neigh != parent:
                dfs(neigh, start, path_len + 1)
        done_visiting[start] = True

    dfs(animal_pos, None, 1)
    return floor(cycle_len/2)

# grid = read_lines("example.txt")
grid = read_lines("input.txt")
print("part one answer", part_one(grid))    